sandbox/lua/decoders/syslog.lua

```diff
@@ -0,0 +1,373 @@
+                 * capture_followed_by('pam_type', '): ')
+wildcard_grammar['pam'] = l.Ct(
+                        (   pam_header
+                          * l.P'session opened for user '
+                          * capture_followed_by('user_name', ' by ')
+                          * capture_followed_by('login_name', '(uid=')
+                          * l.Cg(l.digit^1 / tonumber, 'uid')
+                          * l.P')'
+                        ) + (
+                            pam_header
+                          * l.P'session closed for user '
+                          * l.Cg(l.P(1)^1, 'user_name')
+                        ) + (
+                            pam_header
+                          * l.P'authentication failure; logname='
+                          * capture_followed_by('logname', ' uid=')
+                          * l.Cg(l.digit^1 / tonumber, 'uid')
+                          * l.P' euid='
+                          * l.Cg(l.digit^1 / tonumber, 'euid')
+                          * l.P' tty='
+                          * capture_followed_by('tty', ' ruser=')
+                          * capture_followed_by('ruser', ' rhost=')
+                          * l.Cg((l.P(1) - l.P(' user='))^0, 'rhost')
+                          * (l.P' user=' * l.Cg(l.P(1)^1, 'user'))^-1
+                        ) + (
+                            pam_header
+                          * l.P'check pass; user '
+                          * (l.P'(' * capture_followed_by('user_name', ') '))^-1
+                          * l.P'unknown'
+                        ))
+
+-- end of FIXME
+
+function process_message ()
+    local log = read_message("Payload")
+    local fields = nil
+    if rsyslog_grammar then
+        fields = rsyslog_grammar:match(log)
+        if not fields then return -1 end
+        if fields.timestamp then
+            msg.Timestamp = fields.timestamp
+            fields.timestamp = nil
+        end
+
+        if fields.pri then
+            msg.Severity = fields.pri.severity
+            fields.syslogfacility = fields.pri.facility
+            fields.pri = nil
+        else
+            msg.Severity = fields.syslogseverity or fields["syslogseverity-text"]
+            or fields.syslogpriority or fields["syslogpriority-text"]
+
+            fields.syslogseverity = nil
+            fields["syslogseverity-text"] = nil
+            fields.syslogpriority = nil
+            fields["syslogpriority-text"] = nil
+        end
+
+        if fields.syslogtag then
+            fields.programname = fields.syslogtag.programname
+            msg.Pid = fields.syslogtag.pid
+            fields.syslogtag = nil
+        end
+
+        if not hostname_keep then
+            msg.Hostname = fields.hostname or fields.source
+            fields.hostname = nil
+            fields.source = nil
+        end
+
+        msg.Payload = fields.msg
+        fields.msg = nil
+
+    else
+        msg = decode_message(read_message('raw'))
```

Comment:
I don't understand what your statement about "splitting and decoding phase are tight together" means. Let me step back a bit and explain how Heka works here in more detail:

When a Heka input gets data from the outside world, it is either in native Heka protobuf format or it isn't. If the data *is* in native Heka protobuf format, then it contains an entire Heka message. When the data is decoded, a fully populated Heka message is the result. In this case, Heka will store the raw protobuf data in the pack.MsgBytes attribute. The decoder (usually a ProtobufDecoder) will grab the data from pack.MsgBytes, will deserialize it into a Message struct, and will store that Message struct in the pack.Message attribute, overwriting any message data that may have already been there.

If the data is *not* in native Heka protobuf format, then it is clearly in some other format. Usually this format is a string of some sort. In these cases, Heka creates a mostly-empty Message struct, and stores the data in the Payload field of the Message struct. The decoder must access the message payload to get the raw input. It will then parse this input, and write the extracted data back to the same Message struct. It can populate both base fields and dynamic fields. It may overwrite the original payload, or it may not; many decoders in fact have a `payload_keep` setting that lets you configure whether or not the original payload will be preserved.

Note that in this second case, pack.MsgBytes doesn't come in to the picture at all. If the splitter did not have the `use_message_bytes` option set to true, then the decoder can *not* trust the data that is in pack.MsgBytes. There will be nothing there. All of the useful data will be in the Message struct. If the input has set dynamic message field values, those will be set on the Message struct.

When you call `read_message("Payload")` or `read_message("Fields[foo]")`, you are saying "give me the [Payload|dynamic field 'foo'] value from the `pack.Message` struct". When you call `read_message("raw")`, you are saying "give me the binary protobuf data from the pack.MsgBytes attribute".

Given all of that, here are the things that smell weird about the code that prompted this whole thread:

- All of our existing decoders expect to find their input data *either* in the pack.MsgBytes attribute *or* already in the pack.Message struct. There are no other decoders that sometimes use one, sometimes the other. There might be valid use cases for it, but they're rare, and IMO would need to be defended.
- Your conditional above is based on whether or not an rsyslog_template value was provided and, if so, whether or not it correctly compiled down to an LPEG grammar. Even if you did intend to have a decoder that was expecting its input data in two different locations, that is a horrible way to signal that behaviour.
- Folks usually use syslog for text data. I've never seen anyone use syslog as a transport for binary Heka protobuf data. It's possible, I'm sure, and someone might do it, but AFAIK it hasn't happened yet.

For all of these reasons, it seems to me that `read_message("raw")` isn't the right choice here, and that you don't quite understand what it implies. If I'm wrong about your understanding and it is your explicit intention to be handling binary protobuf data in this decoder, then I still think it's probably a bad idea.

Maybe it would help if you'd explain what behaviour you'd like to see. If the syslog template compiles, then you're parsing the input and distributing the values throughout the message you're emitting. If there's no template, or it doesn't compile, then... what?